README.md
# Kunji NodeJS Library

Setup Authentication and Authorization in your React application in seconds without any verification! Suitable for MVP, side projects or hackathon apps.

## Introduction

`kunji-node` is a Node.js library designed to verify tokens issued by the Kunji authorization server. It comes with an `AuthMiddleware` that can be easily integrated into your Node.js applications to secure routes and endpoints.

You can use the kunji service to integrate login within minutes.

> Also check the frontend react library for client: [kunji-react](https://www.npmjs.com/package/kunji-react)

## Installation

To install `kunji-node`, use npm:

```bash
npm install kunji-node
```
or use yarn:

```bash
yarn add kunji-node
```

## How to register App

This section will be updated after official production release.

## Usage

There are two ways to use the authentication middleware in your application:

1. Using the `Kunji` export to initialize:

```javascript
import { Kunji } from 'kunji-node';

const { AuthMiddleware } = Kunji(appId, publicKey);

// Your express other routes

app.use(AuthMiddleware); // Use the authentication middleware for all routes

// Rest of the code
```

2. Directly importing `AuthMiddleware` with environment variables:

```javascript
import { AuthMiddleware } from 'kunji-node';

// Your express other routes

app.use(AuthMiddleware); // Use the authentication middleware for all routes

// Rest of the code
```

**Note**: The second method requires the environment variables `KUNJI_APP_ID` and `KUNJI_PUBLIC_KEY` to be properly configured. If not configured 500 Response will be given whenever the middleware will be used.

Integrate this middleware into Express before routes requiring authentication. Afterward, you can access the user object using `req.user` in your controller, with the user's unique ID defined as `req.user.uid`.

[![Open in CodeSandbox](https://img.shields.io/badge/Open%20in-CodeSandbox-blue?style=flat-square&logo=codesandbox)](https://codesandbox.io/p/devbox/xjpn87?embed=1)


## Configuration

Make sure to set the following environment variables for configuration:

- `KUNJI_APP_ID`: Your Kunji application ID
- `KUNJI_PUBLIC_KEY`: Your Kunji public key

Example .env (Important: Use double inverted commas for public key when setting in .env):

```
KUNJI_APP_ID=test
KUNJI_PUBLIC_KEY="-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA0QH5YHC2ZmW1w5JYuWeE
YGFU4iXjbDP3HHKO9EiRjWQMiMErHALaHV9WKkgimEMdDmIfPNc6kVgTYTuTTgAa
PL9cdNjJ3qQuPFSR6fx3DL5GUBee99fMQJE0jYhkwO5eyquUMGd1ACss/5bWD1g8
P/saM+Y3BQp2lXse/Z2rsYrgVO1r52iYrqNfkzInj3iS1VrRbIBBnpncSWJPwWap
RdrgZhkitpFOc/jc+wsBZuMKPab7f9o4S8BOJVmO7pG+qY8Bk1kC4OiYyADpErLs
3kOAZZo9IllQJMFZ3C4w7aeTmlZZRchgFhbulHewz7cmN6MtNG6FRE58n9KDFdhY
yQIDAQAB
-----END PUBLIC KEY-----"
```

## Access Token Result (`req.user`)

After the middleware verifies the request, you can access the token result using `req.user`. The interface for the access token result is as follows:

| Field       | Type    | Description                                          |
|-------------|---------|------------------------------------------------------|
| `uid`       | string  | User ID associated with the access token             |
| `role`      | string  | Role assigned to the user                            |
| `iss`       | string  | Issuer of the access token (e.g., Kunji)             |
| `aud`       | string  | Audience for which the access token is intended      |
| `iat`       | number  | Issued at timestamp (UNIX timestamp in seconds)     |
| `exp`       | number  | Expiration timestamp (UNIX timestamp in seconds)    |


## Typescript

When using typescript, use the `AuthRequest` type from the package to be able to access `req.user`. 

Example:

```typescript
import { Response, NextFunction, RequestHandler } from 'express';
import { AuthRequest } from "kunji-node"; 
// Important: when importing middleware directly, Kunji env vars should be configured else it will throw error 500 for every authenticated request

const middleware : RequestHandler = (req: AuthRequest, response: Response, nextFunction : NextFunction) => {
    console.log(req.user.uid)
    // your middleware logic
}
```

## Debugging

1. Pass `debug:true` in `config` object while initialization. 
Example : 
```
const { AuthMiddleware } = Kunji(appId, publicKey, {debug: true});
```

OR

2. Set `KUNJI_ENABLE_DEBUG=true` as an environment variable (.env)

## Contributions

Contributions are welcome! If you find any issues or have suggestions for improvements, feel free to create an issue or submit a pull request.

## License

This project is licensed under the [MIT License](LICENSE).

CHANGELOG.md
# 1.0.0 (2023-12-20)
    - `AuthMiddleware` for using with express to verify tokens



test/index.ts
'use strict';
import express, { Express, Request, Response } from 'express';
import supertest from 'supertest';
import { AuthRequest, Kunji } from '../src/index';
import { expect } from 'chai';

// test application configuration
const appId = 'test';
const publicKey = 'LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUEwUUg1WUhDMlptVzF3NUpZdVdlRQpZR0ZVNGlYamJEUDNISEtPOUVpUmpXUU1pTUVySEFMYUhWOVdLa2dpbUVNZERtSWZQTmM2a1ZnVFlUdVRUZ0FhClBMOWNkTmpKM3FRdVBGU1I2ZngzREw1R1VCZWU5OWZNUUpFMGpZaGt3TzVleXF1VU1HZDFBQ3NzLzViV0QxZzgKUC9zYU0rWTNCUXAybFhzZS9aMnJzWXJnVk8xcjUyaVlycU5ma3pJbmozaVMxVnJSYklCQm5wbmNTV0pQd1dhcApSZHJnWmhraXRwRk9jL2pjK3dzQlp1TUtQYWI3ZjlvNFM4Qk9KVm1PN3BHK3FZOEJrMWtDNE9pWXlBRHBFckxzCjNrT0FaWm85SWxsUUpNRlozQzR3N2FlVG1sWlpSY2hnRmhidWxIZXd6N2NtTjZNdE5HNkZSRTU4bjlLREZkaFkKeVFJREFRQUIKLS0tLS1FTkQgUFVCTElDIEtFWS0tLS0t';
const invalidToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1aWQiOiI2NTdhZDk3N2VkMGVlNjg0MDMzMWNhYjQiLCJyb2xlIjoiTk9STUFMIiwiaWF0IjoxNzAzMTM3NDA2LCJleHAiOjE3MDMxNDEwMDZ9.OCupyvYlAET6EdrviCOHivAZFJNEtdraBpCLu8QM74s';
// This Token has no expiry set for testing purposes
const validToken = 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1aWQiOiI2NTdhZDk3N2VkMGVlNjg0MDMzMWNhYjQiLCJyb2xlIjoiQURNSU4iLCJpc3MiOiJrdW5qaS10ZXN0LWlzc3VlciIsImF1ZCI6InRlc3QiLCJpYXQiOjE3MDMxNDQ0OTR9.nhZxvYVeLGpGqds_IYC-2NuCQpwEfITSHPPOqCBVQAS718QhDPNB5Ps4fXVNkol_2qU64IJdZnPJAYNWVVIz2preyu5TwrwiwJGWglcMGVF-VOC8hlTqbE8sbQd-AGTYV9aVCxZEV__4Ffhfkd74lqYHxH0HA_G6SgSkcnWKWijcptIt0zAc8nqntbfFpdEMR64tTBErVopWWq9xLlG4EB_1ZecIRJ0Sz0gEc4THeW0vR4NJUHR-dWCKE2mMbLLLk20kMMalcjCTd-GII34rLeW_5rkqu3XLvBSOdiK0J_cdkEWIrmTbCvozW8Y_KosjuiCIqZHH-CeJDEKV_H1rmg';

const { AuthMiddleware: authMiddleware } = Kunji(appId, publicKey);

describe('Express Authentication Middleware', () => {
  let app: Express;

  beforeEach(() => {
    app = express();
    app.use(express.json()); // Make sure to parse JSON in request bodies
    app.use(authMiddleware); // Use the authentication middleware for all routes
  });

  it('should pass through authenticated requests', (done) => {
    // Set up a route that requires authentication
    app.get('/authenticated-route', (req: Request, res: Response) => {
      res.sendStatus(200);
    });

    supertest(app)
      .get('/authenticated-route')
      .set('Authorization', `Bearer ${validToken}`)
      .expect(200, done);
  });

  it('should have req.user after successful authenticated request', (done) => {
    // Set up a route that requires authentication
    app.get('/authenticated-route', (req: AuthRequest, res: Response) => {
      try {
        expect(req.user?.uid).to.exist;
        expect(req.user?.uid).to.be.a.string;
        res.sendStatus(200);
      } catch (error) {
        done(error)
      }
    });

    supertest(app)
      .get('/authenticated-route')
      .set('Authorization', `Bearer ${validToken}`)
      .expect(200, done);
  });

  it('should reject unauthenticated requests with a 401 status', (done) => {
    // Set up a route that requires authentication
    app.get('/authenticated-route', (req: Request, res: Response) => {
      res.sendStatus(200);
    });

    supertest(app).get('/authenticated-route').expect(401, done);
  });

  it('should reject requests with invalid tokens with a 401 status', (done) => {

    // Set up a route that requires authentication
    app.get('/authenticated-route', (req: Request, res: Response) => {
      res.sendStatus(200);
    });

    supertest(app)
      .get('/authenticated-route')
      .set('Authorization', `Bearer ${invalidToken}`)
      .expect(401, done);
  });
});

src/middlewares/MiddlewareFactory.ts
import ErrorMessage from '../constants/ErrorMessage'
import { Response, NextFunction } from 'express';
import { AuthRequest, KunjiMiddlewares } from '../types';
import { verifyAccessToken } from '../helpers/TokenHelper';
import { decodeBase64 } from '../helpers';

const MiddlewareFactory = (appId: string, publicKeyBase64: string, config: { debug?: boolean } = { }) : KunjiMiddlewares => {
    const publicKey = decodeBase64(publicKeyBase64) // Decode the Base-64

    // Priority to Config object -> env KUNJI_ENABLE_DEBUG -> default false
    const debug = (log: unknown, isError: boolean = false) => {
        if((config.debug == undefined ? (process.env.KUNJI_ENABLE_DEBUG == 'true' ?? false) : config.debug)){
            // if true
            console.debug(`KUNJI-${(isError ? 'ERROR' : 'LOG')} :`, log)
        }
    }

    return {
            AuthMiddleware : async (req: AuthRequest, res: Response, next: NextFunction) => {

            if(!appId || !publicKey){
                console.error('Kunji Error: AppID and PublicKey is required. Set KUNJI_APP_ID and KUNJI_PUBLIC_KEY or use the default import and initialize there')
                return res.status(500).send(ErrorMessage.STATUS_500_KUNJI_NOT_CONFIGURED);
            }

            const headerToken = req.headers.authorization;
        
            if (!headerToken) {
                return res.status(401).send(ErrorMessage.STATUS_401_NO_TOKEN);
            }
        
            if (headerToken && headerToken.split(' ')[0] !== 'Bearer') {
                return res.status(401).send(ErrorMessage.STATUS_401_NO_TOKEN);
            }
        
            const token = headerToken.split(' ')[1];
        
            try {
                const decodedValue = verifyAccessToken(token, appId, publicKey);
        
                if (decodedValue) {
                    req.user = decodedValue;
                    debug(req.user);
                    return next();
                }
        
                return res.status(401).send(ErrorMessage.STATUS_401_INVALID_TOKEN);
            } catch (error) {
                debug(error)
                return res.status(401).send(ErrorMessage.STATUS_401_INVALID_TOKEN);
            }
        }
    }
}



export default MiddlewareFactory

/* Use Environment Variable */
const service = MiddlewareFactory(process.env.KUNJI_APP_ID ?? '', process.env.KUNJI_PUBLIC_KEY ?? '');

export const AuthMiddleware = service.AuthMiddleware